server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -348,61 +349,68 @@ protected void doInternalExecute(Task task, BulkRequest bulkRequest, Executor ex
+        Set<String> failureStoresToBeRolledOver = new HashSet<>();
+        determineMissingTargets(bulkRequest, indicesToAutoCreate, dataStreamsToBeRolledOver, failureStoresToBeRolledOver);
 
-        // Step 2: filter the list of indices to find those that don't currently exist.
-        final Map<String, IndexNotFoundException> indicesThatCannotBeCreated = new HashMap<>();
-        final ClusterState state = clusterService.state();
-        Map<String, Boolean> indicesToAutoCreate = indices.entrySet()
-            .stream()
-            .filter(entry -> indexNameExpressionResolver.hasIndexAbstraction(entry.getKey(), state) == false)
-            // We should only auto create if we are not requiring it to be an alias
-            .filter(entry -> entry.getValue().isRequireAlias == false)
-            .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().isRequireDataStream));
-
-        // Step 3: Collect all the data streams that need to be rolled over before writing
-        Set<String> dataStreamsToBeRolledOver = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER)
-            ? indices.keySet().stream().filter(target -> {
-                DataStream dataStream = state.metadata().dataStreams().get(target);
-                return dataStream != null && dataStream.rolloverOnWrite();
-            }).collect(Collectors.toSet())
-            : Set.of();
-
-        // Step 4: create all the indices that are missing, if there are any missing. start the bulk after all the creates come back.
         createMissingIndicesAndIndexData(
             task,
             bulkRequest,
             executor,
             listener,
             indicesToAutoCreate,
             dataStreamsToBeRolledOver,
-            indicesThatCannotBeCreated,
+            failureStoresToBeRolledOver,
             startTime
         );
     }
 
-    /*
-     * This method is responsible for creating any missing indices, rolling over a data stream when needed and then
-     *  indexing the data in the BulkRequest
+    private void determineMissingTargets(
+        BulkRequest bulkRequest,
+        Map<String, Boolean> indicesToAutoCreate,
+        Set<String> dataStreamsToBeRolledOver,
+        Set<String> failureStoresToBeRolledOver
+    ) {
+        Map<String, Boolean> indexExistence = new HashMap<>();
+        ClusterState state = clusterService.state();
+        boolean lazyRolloverFeature = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER);
+
+        for (DocWriteRequest<?> request : bulkRequest.requests) {
+            if (request.opType() == OpType.DELETE
+                && request.versionType() != VersionType.EXTERNAL
+                && request.versionType() != VersionType.EXTERNAL_GTE) {
+                continue;
+            }
+            Boolean indexExists = indexExistence.get(request.index());
+            if (indexExists == null) {
```

Comment:
I wanted to avoid using `computeIfAbsent` as I don't think it's very performant when executed a lot (which is the case for the bulk API). I'll update this to use `Map.containsKey`.

Reply:
Sorry maybe performant isn't the right word. I was thinking of https://github.com/elastic/elasticsearch/commit/99e72b02ba46038d898e981f9c39d07a392735f8, which I remembered as "the lambda argument of `computeIfAbsent` can take up significant space on the heap when invoked often". After reading the commit description again, I see it's a bit more nuanced than that. But I'm inclined to think that that statement also applies to this situation (as this will be called for every request in every bulk request). What do you think?

Reply:
I'm okay with extracting the lambda to outside the loop. That would still create a new lambda for every bulk request, but that's fine, I think -- extracting it to a class method or a static class field is overkill IMO. I think readability is still fine with this approach.

Also, technically, I guess `computeIfAbsent` still makes use of a tri-state boolean, it's just hidden away from us :)

Reply:
I am ok with the `contains` too I was just as surprised as @jbaiera that's why I asked.